Upnp properties

dgomes opened the pull request:
This pull request improves the UPnP component with IGD sensors containing bytes sent/received and packets sent/received

dgomes committed:
make port mapping optional

dgomes committed:
dependencies + improvements

dgomes committed:
Added bytes and packets sensors from IGD

dgomes committed:
flake8 check

dgomes committed:
new sensor with upnp counters

dgomes committed:
checks

dgomes committed:
Merge remote-tracking branch 'home-assistant/dev' into upnp_properties

dgomes committed:
whitespaces in blank line

fabaff commented:
The checklist of the PR template covers a bunch of topics which are missing in the PR.

dgomes commented:
Where can I find the template ? More than willing to cover whats needed

dgomes committed:
Merge remote-tracking branch 'home-assistant/dev' into upnp_properties

dgomes committed:
requirements update

dgomes committed:
added sensor.upnp to .coveragerc

dgomes committed:
downgrade miniupnpc

Latest version of miniupnpc is 2.0, but pypi only has 1.9

Fortunately it is enough

balloob commented on homeassistant/components/upnp.py:
```diff
@@ -1,32 +1,54 @@
+    hass.data[DATA_UPNP] = upnp
 
     upnp.discoverdelay = 200
     upnp.discover()
```
You cannot do I/O inside a coroutine.

dgomes replied:
I understood hass.data would be a global dictionary and wouldn't involve any I/O

But I'm moving everything back to setup() and update()

balloob replied:
Assigning to data doesn't do I/O. This comment was about `upnp.discover()` call. Since you changed it back to no longer be a coroutine this comment can be ignored.

balloob commented on homeassistant/components/upnp.py:
```diff
@@ -36,17 +58,22 @@ def setup(hass, config):
-    host = base_url.hostname
-    external_port = internal_port = base_url.port
+    unit = config[DOMAIN].get(CONF_UNITS)
+    discovery.load_platform(hass, 'sensor', DOMAIN, {'unit': unit}, config)
```
You cannot call sync methods inside a coroutine.

I would suggest you change it back to `setup` and not make it async.

dgomes replied:
will do

balloob commented on homeassistant/components/upnp.py:
```diff
@@ -36,17 +58,22 @@ def setup(hass, config):
+    discovery.load_platform(hass, 'sensor', DOMAIN, {'unit': unit}, config)
+
+    port_mapping = config[DOMAIN].get(CONF_ENABLE_PORT_MAPPING)
+    if port_mapping:
```
Make it a guard clause:

```python
if not port_mapping:
    return True

base_url = …
```

dgomes replied:
check

balloob commented on homeassistant/components/sensor/upnp.py:
```diff
@@ -0,0 +1,78 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+""" sensor_type: [friendly_name, convert_unit, icon] """
```
Use `#` for comments. Only use `"""` for first line in module/class/method to indicate it's a doc string.

dgomes replied:
check

balloob commented on homeassistant/components/sensor/upnp.py:
```diff
@@ -0,0 +1,78 @@
+    def async_update(self):
+        """Get the latest information from the IGD."""
+        if self.type == "byte_received":
+            self._state = self._upnp.totalbytereceived()
```
Do these methods do I/O ? If so, this shouldn't be `async_update` but instead `update`

dgomes replied:
check

balloob commented on homeassistant/components/upnp.py:
```diff
@@ -1,32 +1,54 @@
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers import discovery
+
+REQUIREMENTS = ['miniupnpc==1.9']
```
Will this install miniupnp ?

balloob reviewed (changes requested):
You've updated a few places to be async instead of the sync version. However, it seems that there is still I/O being done in those methods. Anything that hits a network interface or talks to disk/network cannot be run inside a coroutine.

fabaff commented:
Documentation PR: #2832

dgomes committed:
revert to non async

miniupnpc will do network calls, so this component can’t be moved to
coroutine

dgomes committed:
hof hof

forgot to remove import ot asyncio

dgomes reviewed (commented):
I've addressed all the changes

balloob commented:
Awesome, thanks!

dgomes commented:
Thanks for the mentoring, looking forward to more contributions.
